topologies/binding/options.go

```diff
@@ -101,43 +101,48 @@ func (d *dialer) loadCertificates() (*x509.CertPool, tls.Certificate, error) {
 //
 //lint:ignore U1000 will be used by the binding.
 func (d *dialer) dialGRPC(ctx context.Context, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	switch {
-	case d.Insecure:
-		tc := insecure.NewCredentials()
-		opts = append(opts, grpc.WithTransportCredentials(tc))
-	case d.SkipVerify:
-		tc := credentials.NewTLS(&tls.Config{InsecureSkipVerify: true})
-		opts = append(opts, grpc.WithTransportCredentials(tc))
-	case d.MutualTls:
-		trusBundle, keyPair, err := d.loadCertificates()
-		if err != nil {
-			return nil, err
+	if len(opts) == 0 {
+		switch {
+		case d.Insecure:
+			tc := insecure.NewCredentials()
+			opts = append(opts, grpc.WithTransportCredentials(tc))
+		case d.SkipVerify:
+			tc := credentials.NewTLS(&tls.Config{InsecureSkipVerify: true})
+			opts = append(opts, grpc.WithTransportCredentials(tc))
+		case d.MutualTls:
+			trusBundle, keyPair, err := d.loadCertificates()
+			if err != nil {
+				return nil, err
+			}
+			tls := &tls.Config{
+				Certificates: []tls.Certificate{keyPair},
+				RootCAs:      trusBundle,
+			}
+			tlsConfig := credentials.NewTLS(tls)
+			opts = append(opts, grpc.WithTransportCredentials(tlsConfig))
 		}
-		tls := &tls.Config{
-			Certificates: []tls.Certificate{keyPair},
-			RootCAs:      trusBundle,
+		if d.Username != "" {
+			c := &creds{d.Username, d.Password, !d.Insecure}
+			opts = append(opts, grpc.WithPerRPCCredentials(c))
+		}
+		if d.MaxRecvMsgSize != 0 {
+			opts = append(opts, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(int(d.MaxRecvMsgSize))))
+		}
+		if d.Timeout == 0 {
+			retryOpt := grpc_retry.WithPerRetryTimeout(time.Duration(d.Timeout) * time.Second)
+			opts = append(opts,
+				grpc.WithStreamInterceptor(grpc_retry.StreamClientInterceptor(retryOpt)),
+				grpc.WithUnaryInterceptor(grpc_retry.UnaryClientInterceptor(retryOpt)),
+			)
 		}
-		tlsConfig := credentials.NewTLS(tls)
-		opts = append(opts, grpc.WithTransportCredentials(tlsConfig))
-	}
-	if d.Username != "" {
-		c := &creds{d.Username, d.Password, !d.Insecure}
-		opts = append(opts, grpc.WithPerRPCCredentials(c))
-	}
-	if d.MaxRecvMsgSize != 0 {
-		opts = append(opts, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(int(d.MaxRecvMsgSize))))
 	}
 	if d.Timeout == 0 {
 		return grpc.DialContext(ctx, d.Target, opts...)
+	} else {
+		ctx, cancelFunc := context.WithTimeout(ctx, time.Duration(d.Timeout)*time.Second)
+		defer cancelFunc()
+		return grpc.DialContext(ctx, d.Target, opts...)
```

Comment:
let's rework this logic so that there is a single grpc.DialContext call:

```
if d.Timeout > 0 {
  var cancel context.CancelFunc
  ctx, cancel = context.WithTimeout(ctx, d.Timeout * time.Second)
  defer cancel()
}
return grpc.DialContext(ctx, d.Target, opts...)
```

Reply:
done